## Connecting to the Database

In [27]:
pip install psycopg2-binary;

Note: you may need to restart the kernel to use updated packages.


In [28]:
pip install mxnet

Note: you may need to restart the kernel to use updated packages.


In [29]:
import psycopg2

In [30]:
# Put the password in 
PGHOST = 'tripdatabase2.cmaaautpgbsf.us-east-2.rds.amazonaws.com'
PGDATABASE = ''
PGUSER = 'postgres'
PGPASSWORD = 'Josh1234'

In [31]:
# Database Context Manager
try:   
    # Set up a connection to the postgres server.    
    conn = psycopg2.connect(user = PGUSER,
                            port = "5432",
                            password = PGPASSWORD,
                            host = PGHOST,
                            database = PGDATABASE)
    # Create a cursor object
    cursor = conn.cursor()   
    cursor.execute("SELECT version();")
    record = cursor.fetchone()
    print("Connection Success:", record,"\n")

except (Exception, psycopg2.Error) as error:
    print("Error while connecting to PostgreSQL", error)

Connection Success: ('PostgreSQL 12.5 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit',) 



In [32]:
import pandas as pd
import numpy as np
import Queries
import matplotlib.pyplot as plt
import seaborn as sns
import importlib
import s3fs
importlib.reload(Queries)

<module 'Queries' from '/root/Citi-Bike-Expansion/Queries.py'>

In [33]:
# The S3 Bucket that will be used to store the data should be created beforehand
ACCESS_KEY_ID = 'AKIARJEUISD2VILSZ6HM'
ACCESS_SECRET_KEY = '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'

fs = s3fs.S3FileSystem(anon=False, key = ACCESS_KEY_ID, secret= ACCESS_SECRET_KEY)

## Initial Data Import

The key data file that is needed to do the machine learning contains data for every zipcode in the United States. The aim is to come up with a universal classification model that can be used for any location in the States. Therefore, a couple of columns are going to be dropped from the start:
- Country Name 
- State Name & State Code

Some other columns that will be dropped are:
- Cities Postal Code: These codes are not the same as the zipcodes that we have).
- Area in Miles: We will continue to work in meters.
- Total Housing: There are two identical columns, we drop one.
- Predicted Qualified Leads: Not actually a zipcode property.

In [34]:
zipcode_file = "s3://williams-citibike/zipcodes.csv"

with fs.open(zipcode_file, 'r') as file:
    zipcodes = pd.read_csv(file, sep=',', low_memory = False)
    
zipcodes = zipcodes[zipcodes.country_name == 'United States']
zipcodes.drop(columns = ['cities_postalcode_id', 'state_name', 'state_code', 
                         'country_name','area_land_sq_miles', 'area_water_sq_miles',
                         'units_in_structure_housing_units_total_housing_units',
                         'predicted_qualified_leads_per_month'], inplace=True)

In [35]:
column_renames = ['zipcode', 'cbsa', 'ziptype', 'land_area_sqm',
                  'water_area_sqm', 'total_population', 'total_population_18_over',
                  'median_age', 'pct_labor_force_16_over', 
                  'pct_labor_force_unemployed_16_over', 'pct_armed_force_16_over', 
                  'pct_labor_force_employed_16_over', 'median_household_income',
                  'family_income_per_capita', 'total_housing_units',
                  'pct_vacant_housing', 'occupied_housing_units',
                  'pct_no_vehicle_of_occupied_housing', 'median_price_owner_occupied_units',
                  'median_rent_occupied_units_paying_rent', 'pct_1_unit_attached',
                  'pct_1_unit_detached', 'pct_2_units', 'pct_3-4_units', 'pct_5-9_units',
                  'pct_10-19_units', 'pct_20+_units', 'avg_household_size', 'pct_bachelors+_25_over',
                  'pct_diff_housing_from_last_year', 'pct_same_housing_from_last_year',
                  'resident_since_last_year', 'population_in_college_grad', 'population_density', 
                  'pct_population_in_college_grad', 'zipcode_segment'
                 ]

zipcodes.columns = column_renames

Core-Based Statistical Area (CBSA): "A CBSA is one ore more adjacent counties or county equivalents that have at least one ubran core area of at least 10,000 population, plus adjacent territory that has a high degree of social and economic integration with the core as measured by community ties". There are 927 CBSAs in the United States, each containing multiple zipcodes, categorized into two types: 
- Metropolitan: one urbanized area of 50,000 or more inhabitants.
- Micropolitan: one urban cluster of at least 10,000 but less than 50,000 population.

**In the dataset NaN values in this column mean that the zipcode isn't within a CBSA and we will fill the column with 'NO CBSA'.**

https://www.census.gov/programs-surveys/metro-micro/about.html#:~:text=Each%20metropolitan%20statistical%20area%20must,but%20less%20than%2050%2C000%20population
https://en.wikipedia.org/wiki/Core-based_statistical_area

In [37]:
zipcodes.cbsa = zipcodes.cbsa.fillna('NO CBSA')

Zip Type: There are two types of zip codes
- Zip Code Areas are typical ZIP Codes without any special designation. 
- Post Office or large volume customers typically occupy a very small geography and have few residential addresses located within them. 

https://udsmapper.org/zip-code-to-zcta-crosswalk/

Zip Code Segment (Categorical):
Describes the density and the type of people that are working in the area. This segmentation could be really useful, but it is missing for 93% of the data so it will be dropped.

In [52]:
zipcodes = zipcodes.drop(columns=['zipcode_segment'])

### Handling Missing Data

In [38]:
zipcodes.isna().sum()

zipcode                                       0
cbsa                                          0
ziptype                                    2747
land_area_sqm                              2770
water_area_sqm                             2770
total_population                           2770
total_population_18_over                   2770
median_age                                 3377
pct_labor_force_16_over                    2770
pct_labor_force_unemployed_16_over         2770
pct_armed_force_16_over                    2770
pct_labor_force_employed_16_over           2770
median_household_income                    4900
family_income_per_capita                   3457
total_housing_units                        2770
pct_vacant_housing                         2770
occupied_housing_units                     2770
pct_no_vehicle_of_occupied_housing         2770
median_price_owner_occupied_units          5414
median_rent_occupied_units_paying_rent     8673
pct_1_unit_attached                     

There are a lot of columns that have a missing value count of 2770, which tells me that there is a set of rows where all of those columns are blank. I will use the land_area_sqm column to try to find them, because regardless of the zipcode it should have land. 

In [40]:
zipcodes[zipcodes.land_area_sqm.isna()]

zipcode     cbsa                               ziptype  land_area_sqm  \
0       00210  NO CBSA                                   NaN            NaN   
1       00211  NO CBSA                                   NaN            NaN   
2       00212  NO CBSA                                   NaN            NaN   
3       00213  NO CBSA                                   NaN            NaN   
4       00214  NO CBSA                                   NaN            NaN   
...       ...      ...                                   ...            ...   
43268   99332  NO CBSA                                   NaN            NaN   
43308   99512  NO CBSA                                   NaN            NaN   
43359   99584  NO CBSA                                   NaN            NaN   
43435   99675  NO CBSA  Post Office or large volume customer            NaN   
43528   99779  NO CBSA                                   NaN            NaN   

       water_area_sqm  total_population  total_population_18_over  median_age  \
0                 NaN               NaN                       NaN         NaN   
1                 NaN               NaN                       NaN         NaN   
2                 NaN               NaN                       NaN         NaN   
3                 NaN               NaN                       NaN         NaN   
4                 NaN               NaN                       NaN         NaN   
...               ...               ...                       ...         ...   
43268             NaN               NaN                       NaN         NaN   
43308             NaN               NaN                       NaN         NaN   
43359             NaN               NaN                       NaN         NaN   
43435             NaN               NaN                       NaN         NaN   
43528             NaN               NaN                       NaN         NaN   

       pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
0                          NaN                                 NaN  ...   
1                          NaN                                 NaN  ...   
2                          NaN                                 NaN  ...   
3                          NaN                                 NaN  ...   
4                          NaN                                 NaN  ...   
...                        ...                                 ...  ...   
43268                      NaN                                 NaN  ...   
43308                      NaN                                 NaN  ...   
43359                      NaN                                 NaN  ...   
43435                      NaN                                 NaN  ...   
43528                      NaN                                 NaN  ...   

       pct_20+_units  avg_household_size  pct_bachelors+_25_over  \
0                NaN                 NaN                     NaN   
1                NaN                 NaN                     NaN   
2                NaN                 NaN                     NaN   
3                NaN                 NaN                     NaN   
4                NaN                 NaN                     NaN   
...              ...                 ...                     ...   
43268            NaN                 NaN                     NaN   
43308            NaN                 NaN                     NaN   
43359            NaN                 NaN                     NaN   
43435            NaN                 NaN                     NaN   
43528            NaN                 NaN                     NaN   

       pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
0                                  NaN                              NaN   
1                                  NaN                              NaN   
2                                  NaN                              NaN   
3                                  NaN                            

Let's check to make sure that all of those values are missing in each column of each row.

In [44]:
(zipcodes[zipcodes.land_area_sqm.isna()].isna().sum(axis=1)>30).sum()

2770

In [49]:
zipcodes = zipcodes.drop(zipcodes[zipcodes.land_area_sqm.isna()].index).reset_index(drop=True)

In [53]:
zipcodes.isna().sum()

zipcode                                      0
cbsa                                         0
ziptype                                      0
land_area_sqm                                0
water_area_sqm                               0
total_population                             0
total_population_18_over                     0
median_age                                 607
pct_labor_force_16_over                      0
pct_labor_force_unemployed_16_over           0
pct_armed_force_16_over                      0
pct_labor_force_employed_16_over             0
median_household_income                   2130
family_income_per_capita                   687
total_housing_units                          0
pct_vacant_housing                           0
occupied_housing_units                       0
pct_no_vehicle_of_occupied_housing           0
median_price_owner_occupied_units         2644
median_rent_occupied_units_paying_rent    5903
pct_1_unit_attached                          0
pct_1_unit_de

The percent of population in college or graduate school column is calculated using the population in college or graduate shcool and the total population. In the dataset if the number of people in college is 0, then the percentage column is NaN. So we will fill those with zeros.

In [77]:
zipcodes.pct_population_in_college_grad = zipcodes.pct_population_in_college_grad.fillna(0)

In [15]:
station_zipcodes = Queries.get_zipcode_stations(conn)
station_zipcodes.zipcode = station_zipcodes.astype('str')
station_zipcodes.zipcode = station_zipcodes.zipcode.str.zfill(5)

In [16]:
station_zipcodes.head()

zipcode  num_stations
0   000-1             1
1   02108             4
2   02109             2
3   02110            10
4   02111             5

In [17]:
bikeshare_zipcodes = zipcodes.merge(station_zipcodes, how='inner', on='zipcode')

In [20]:
bikeshare_zipcodes.head()

zipcode                           cbsa        ziptype  land_area_sqm  \
0   02108  Metropolitan Statistical Area  Zip Code Area       353039.0   
1   02109  Metropolitan Statistical Area  Zip Code Area       451168.0   
2   02110  Metropolitan Statistical Area  Zip Code Area       477072.0   
3   02111  Metropolitan Statistical Area  Zip Code Area       699955.0   
4   02113  Metropolitan Statistical Area  Zip Code Area       259756.0   

   water_area_sqm  total_population  total_population_18_over  median_age  \
0             0.0            4049.0                    3622.0        35.2   
1        292691.0            4015.0                    3713.0        32.9   
2        234496.0            2124.0                    2012.0        53.6   
3         43757.0            7564.0                    6790.0        32.9   
4             0.0            7495.0                    7188.0        30.2   

   pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
0                    0.699                               0.077  ...   
1                    0.749                               0.053  ...   
2                    0.670                               0.018  ...   
3                    0.624                               0.088  ...   
4                    0.824                               0.020  ...   

   avg_household_size  pct_bachelors+_25_over  \
0                1.70                   0.829   
1                1.72                   0.736   
2                1.61                   0.874   
3                2.00                   0.526   
4                1.72                   0.772   

   pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
0                            0.263                            0.728   
1                            0.324                            0.625   
2                            0.176                            0.806   
3                            0.203                            0.764   
4                            0.309                            0.680   

   resident_since_last_year  population_in_college_grad  population_density  \
0                    4049.0                       580.0        29772.058824   
1                    3978.0                       235.0        23074.712644   
2                    2124.0                       197.0        11543.478261   
3                    7493.0                      2069.0        28014.814815   
4                    7495.0                       976.0        74950.000000   

   pct_population_in_college_grad                  zipcode_segment  \
0                        0.143245     Medium density professionals   
1                        0.058531     Medium density professionals   
2                        0.092750        Low density professionals   
3                        0.273533                 College students   
4                        0.130220  Very high density professionals   

   num_stations  
0             4  
1             2  
2            10  
3             5  
4             2  

[5 rows x 37 columns]



***Observation: All the zipcodes that have bikeshare stations are metropolitan statistical areas. (See Commented Code Below)***



In [16]:
# bikeshare_zipcodes.cbsa.unique()

In [17]:
bikeshare_zipcodes.shape

(318, 37)

In [18]:
bikeshare_zipcodes.isna().sum()

zipcode                                    0
cbsa                                       1
ziptype                                    0
land_area_sqm                              0
water_area_sqm                             0
total_population                           0
total_population_18_over                   0
median_age                                13
pct_labor_force_16_over                    0
pct_labor_force_unemployed_16_over         0
pct_armed_force_16_over                    0
pct_labor_force_employed_16_over           0
median_household_income                   16
family_income_per_capita                  13
total_housing_units                        0
pct_vacant_housing                         0
occupied_housing_units                     0
pct_no_vehicle_of_occupied_housing         0
median_price_owner_occupied_units         18
median_rent_occupied_units_paying_rent    15
pct_1_unit_attached                        0
pct_1_unit_detached                        0
pct_2_unit

**Median Age**

In [19]:
bikeshare_zipcodes[bikeshare_zipcodes.median_age.isna()]

zipcode                           cbsa  \
37    02203  Metropolitan Statistical Area   
66    10020  Metropolitan Statistical Area   
90    10168  Metropolitan Statistical Area   
91    10173  Metropolitan Statistical Area   
92    10199  Metropolitan Statistical Area   
155   20240  Metropolitan Statistical Area   
157   20245  Metropolitan Statistical Area   
160   20418  Metropolitan Statistical Area   
164   20551  Metropolitan Statistical Area   
166   20560  Metropolitan Statistical Area   
167   20565  Metropolitan Statistical Area   
168   20566  Metropolitan Statistical Area   
170   20593  Metropolitan Statistical Area   

                                  ziptype  land_area_sqm  water_area_sqm  \
37   Post Office or large volume customer        82171.0             0.0   
66                          Zip Code Area        71560.0             0.0   
90   Post Office or large volume customer        13211.0             0.0   
91   Post Office or large volume customer        12297.0             0.0   
92   Post Office or large volume customer        65414.0             0.0   
155  Post Office or large volume customer        31673.0             0.0   
157  Post Office or large volume customer        64474.0             0.0   
160  Post Office or large volume customer        29418.0             0.0   
164  Post Office or large volume customer        23513.0             0.0   
166  Post Office or large volume customer        53607.0             0.0   
167  Post Office or large volume customer       173126.0             0.0   
168  Post Office or large volume customer       105615.0             0.0   
170  Post Office or large volume customer        77132.0             0.0   

     total_population  total_population_18_over  median_age  \
37                0.0                       0.0         NaN   
66                0.0                       0.0         NaN   
90                0.0                       0.0         NaN   
91                0.0                       0.0         NaN   
92                0.0                       0.0         NaN   
155               0.0                       0.0         NaN   
157               0.0                       0.0         NaN   
160               0.0                       0.0         NaN   
164               0.0                       0.0         NaN   
166               0.0                       0.0         NaN   
167               0.0                       0.0         NaN   
168               0.0                       0.0         NaN   
170               0.0                       0.0         NaN   

     pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
37                       0.0                                 0.0  ...   
66                       0.0                                 0.0  ...   
90                       0.0                                 0.0  ...   
91                       0.0                                 0.0  ...   
92                       0.0                                 0.0  ...   
155                      0.0                                 0.0  ...   
157                      0.0                                 0.0  ...   
160                      0.0                                 0.0  ...   
164                      0.0                                 0.0  ...   
166                      0.0                                 0.0  ...   
167                      0.0                                 0.0  ...   
168                      0.0                                 0.0  ...   
170                      0.0                                 0.0  ...   

     avg_household_size  pct_bachelors+_25_over  \
37                  NaN                     0.0   
66                  NaN                     0.0   
90                  NaN                     0.0   
91                  NaN                     0.0   
92                  NaN                     0.0   
155                 NaN                     0.0   
157                 NaN        

Zipcodes that have missing data for the median has missing data for all the columns. Unfortunately, this data doesn't exist anywhere.

In [20]:
bikeshare_zipcodes = bikeshare_zipcodes.drop(bikeshare_zipcodes[bikeshare_zipcodes.median_age.isna()].index)

In [21]:
bikeshare_zipcodes.isna().sum()

zipcode                                    0
cbsa                                       1
ziptype                                    0
land_area_sqm                              0
water_area_sqm                             0
total_population                           0
total_population_18_over                   0
median_age                                 0
pct_labor_force_16_over                    0
pct_labor_force_unemployed_16_over         0
pct_armed_force_16_over                    0
pct_labor_force_employed_16_over           0
median_household_income                    3
family_income_per_capita                   0
total_housing_units                        0
pct_vacant_housing                         0
occupied_housing_units                     0
pct_no_vehicle_of_occupied_housing         0
median_price_owner_occupied_units          5
median_rent_occupied_units_paying_rent     2
pct_1_unit_attached                        0
pct_1_unit_detached                        0
pct_2_unit

**Median Household Income**

In [22]:
bikeshare_zipcodes[bikeshare_zipcodes.median_household_income.isna()]

zipcode                           cbsa        ziptype  land_area_sqm  \
94    10282  Metropolitan Statistical Area  Zip Code Area       179389.0   
152   20052  Metropolitan Statistical Area  Zip Code Area         7062.0   
309   94720  Metropolitan Statistical Area  Zip Code Area      1610594.0   

     water_area_sqm  total_population  total_population_18_over  median_age  \
94          55392.0            5635.0                    4036.0        34.5   
152             0.0             431.0                     431.0        19.8   
309             0.0            3166.0                    3048.0        19.7   

     pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
94                     0.781                               0.022  ...   
152                    0.357                               0.201  ...   
309                    0.307                               0.233  ...   

     avg_household_size  pct_bachelors+_25_over  \
94                 2.32                   0.857   
152                 NaN                   0.852   
309                3.17                   0.889   

     pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
94                             0.171                            0.785   
152                            0.619                            0.281   
309                            0.609                            0.211   

     resident_since_last_year  population_in_college_grad  population_density  \
94                     5419.0                       165.0        81666.666667   
152                     431.0                       423.0       143666.666667   
309                    3166.0                      3094.0         5090.032154   

     pct_population_in_college_grad             zipcode_segment  num_stations  
94                         0.029281  High density professionals             2  
152                        0.981439            College students             2  
309                        0.977258            College students             2  

[3 rows x 37 columns]

Median Household Income for **10282** from 1st source 250,000 and was confirmed by 2nd Source. No data for **20052** in any source. No Median Income but, there is Average Household Income for **94720** in 2nd source at 319,462.

First Source
https://www.census.gov/acs/www/data/data-tables-and-tools/narrative-profiles/2019/

Second Check
https://www.incomebyzipcode.com/

In [23]:
bikeshare_zipcodes.loc[94,'median_household_income'] = 250000
bikeshare_zipcodes.loc[309, 'median_household_income'] = 319462

**Median Price of Owner Occupied Housing**

In [24]:
bikeshare_zipcodes[bikeshare_zipcodes.median_price_owner_occupied_units.isna()]

zipcode                           cbsa        ziptype  land_area_sqm  \
34    02163  Metropolitan Statistical Area  Zip Code Area       256728.0   
36    02199  Metropolitan Statistical Area  Zip Code Area       148225.0   
60    10013  Metropolitan Statistical Area  Zip Code Area      1425085.0   
152   20052  Metropolitan Statistical Area  Zip Code Area         7062.0   
309   94720  Metropolitan Statistical Area  Zip Code Area      1610594.0   

     water_area_sqm  total_population  total_population_18_over  median_age  \
34          32166.0            2575.0                    2451.0        27.4   
36              0.0            1260.0                    1218.0        53.4   
60              0.0           28211.0                   23643.0        36.9   
152             0.0             431.0                     431.0        19.8   
309             0.0            3166.0                    3048.0        19.7   

     pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
34                     0.242                               0.100  ...   
36                     0.580                               0.048  ...   
60                     0.672                               0.044  ...   
152                    0.357                               0.201  ...   
309                    0.307                               0.233  ...   

     avg_household_size  pct_bachelors+_25_over  \
34                 2.02                   0.980   
36                 1.55                   0.868   
60                 2.20                   0.627   
152                 NaN                   0.852   
309                3.17                   0.889   

     pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
34                             0.391                            0.468   
36                             0.357                            0.628   
60                             0.134                            0.844   
152                            0.619                            0.281   
309                            0.609                            0.211   

     resident_since_last_year  population_in_college_grad  population_density  \
34                     2562.0                      2150.0        26010.101010   
36                     1243.0                        71.0        22105.263158   
60                    27842.0                      2507.0        51292.727273   
152                     431.0                       423.0       143666.666667   
309                    3166.0                      3094.0         5090.032154   

     pct_population_in_college_grad               zipcode_segment  \
34                         0.834951              College students   
36                         0.056349  Medium density professionals   
60                         0.088866    High density professionals   
152                        0.981439              College students   
309                        0.977258              College students   

     num_stations  
34              2  
36              2  
60             21  
152             2  
309             2  

[5 rows x 37 columns]

For both the NYC 10013 and the California 94720 zipcode the median owner occupied price was more than 2,000,000. 

*Data came from the 1st source listed above**

In [25]:
bikeshare_zipcodes.loc[60,'median_price_owner_occupied_units'] = 2000000
bikeshare_zipcodes.loc[309,'median_price_owner_occupied_units'] = 2000000

**Median Rent of Occupied Housing Paying Rent** - No data could be found to be inputted

**Average Household Size** - No data could be found to be inputted

**Zipcode Segment** - Most of the data in the regular df doesn't have this data (93%), which is a problem if the model is built on top of it, we will remove this column.


In [26]:
bikeshare_zipcodes = bikeshare_zipcodes.drop(columns=['zipcode_segment'])

In [27]:
bikeshare_zipcodes = bikeshare_zipcodes.reset_index(drop=True)

### Categorical Data Encoding

## The Machine Learning

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [29]:
X = bikeshare_zipcodes.drop(columns = ['num_stations'])
y = bikeshare_zipcodes.num_stations

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

In [31]:
dummy_regr = DummyRegressor(strategy='median')
dummy_regr.fit(X_train, y_train)

y_pred = dummy_regr.predict(X_test)


In [32]:
mean_squared_error(y_test, y_pred, squared=False)

10.870282585023302